extensions/azure-extensions/src/main/java/io/druid/firehose/azure/StaticAzureBlobStoreFirehoseFactory.java

```diff
@@ -0,0 +1,121 @@
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.common.base.Charsets;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Throwables;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Lists;
+import com.metamx.common.CompressionUtils;
+import com.metamx.common.logger.Logger;
+import io.druid.data.input.Firehose;
+import io.druid.data.input.FirehoseFactory;
+import io.druid.data.input.impl.FileIteratingFirehose;
+import io.druid.data.input.impl.StringInputRowParser;
+import io.druid.storage.azure.AzureByteSource;
+import io.druid.storage.azure.AzureStorage;
+import org.apache.commons.io.IOUtils;
+import org.apache.commons.io.LineIterator;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.net.URI;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+
+public class StaticAzureBlobStoreFirehoseFactory implements FirehoseFactory<StringInputRowParser> {
+  private static final Logger log = new Logger(StaticAzureBlobStoreFirehoseFactory.class);
+
+  private final AzureStorage azureStorage;
+  private final List<AzureBlob> blobs;
+
+  @JsonCreator
+  public StaticAzureBlobStoreFirehoseFactory(
+      @JacksonInject("azureStorage") AzureStorage azureStorage,
+      @JsonProperty("blobs") AzureBlob[] blobs
+  )
+  {
+    this.azureStorage = azureStorage;
+    this.blobs = ImmutableList.copyOf(blobs);
+  }
+
+  @JsonProperty
+  public List<AzureBlob> getBlobs() { return blobs; }
+
+  @Override
+  public Firehose connect(StringInputRowParser stringInputRowParser) throws IOException {
+    Preconditions.checkNotNull(azureStorage, "null azureStorage");
+
+    final LinkedList<AzureBlob> objectQueue = Lists.newLinkedList(blobs);
+
+    return new FileIteratingFirehose(
+        new Iterator<LineIterator>() {
+          @Override
+          public boolean hasNext() {
+            return !objectQueue.isEmpty();
+          }
+
+          @Override
+          public LineIterator next() {
+            final AzureBlob nextURI = objectQueue.poll();
+
+            final String container = nextURI.getContainer();
+            final String path = nextURI.getPath().startsWith("/")
+                ? nextURI.getPath().substring(1)
+                : nextURI.getPath();
+
+            try {
+              final InputStream innerInputStream = new AzureByteSource(azureStorage, container, path).openStream();
+
+
+              final InputStream outerInputStream = path.endsWith(".gz")
+                  ? CompressionUtils.gzipInputStream(innerInputStream)
+                  : innerInputStream;
+
+              return IOUtils.lineIterator(
+                  new BufferedReader(
+                      new InputStreamReader(outerInputStream, Charsets.UTF_8)
+                  )
+              );
+            } catch (Exception e) {
+              log.error(e,
```

Comment:
I honestly don't know, haven't dealt with Java streams enough to be able to make that call myself.

From some basic source code digging, it seems the gzipInputStream will throw IOException if parent stream (in this case the stream from blob store) is closed. Since close appears to be idempotent here, I'll close them both, starting with outerInputStream.

close however, can also throw IOException, and I plan on just letting that bubble up.